### PR TITLE
feat: update model

### PR DESCRIPTION
**Pull request by borisdayma**
Few model updates:

* projection layer not applied when using convolutions
* position embeddings set as parameter

Feel free to merge.

### CODE REVIEW

**Comment by borisdayma** [comment]
Questions:

* for the projection to codebook layer, should we add any non-linearity before so that it's a more useful layer? I added one so it doesn't feel like a waste
* ...

File: vit_vqgan/modeling_vit_vqgan.py [hunk 1]
```diff
@@ -299,41 +299,30 @@ class VitEncoder(nn.Module):
...
+        hidden_states = nn.Dropout(rate=self.config.dropout)(hidden_states, deterministic=deterministic)
+        hidden_states = Transformer(self.config, dtype=self.dtype)(hidden_states, deterministic=deterministic)
         return hidden_states
 
 
 class VitDecoder(nn.Module):
     config: ViTVQConfig
     dtype: jnp.dtype = jnp.float32
 
+    @nn.compact
```
Comment (patil-suraj) [inline vit_vqgan/modeling_vit_vqgan.py:333]: Think we can delete `setup` here since we are using `nn.compact` with call now.

File: vit_vqgan/modeling_vit_vqgan.py [hunk 2]
```diff
@@ -345,8 +334,15 @@ def setup(self):
...
         self.transformer = Transformer(self.config, dtype=self.dtype)
 
+    @nn.compact
     def __call__(self, hidden_states, deterministic: bool = True):
-        hidden_states = hidden_states + self.position_embeddings(self.position_ids)
+        assert hidden_states.ndim == 3
+        position_embeddings = self.param(
+            "pos_embedding",
+            jax.nn.initializers.normal(self.config.initializer_range),
+            (1, 1, hidden_states.shape[1], hidden_states.shape[2]),
```
Comment (patil-suraj) [inline vit_vqgan/modeling_vit_vqgan.py]: The shape doesn't look right here, think it should be `(1, num_patches,  self.config.hidden_size)`. 
`(1, hidden_states.shape[1], hidden_states.shape[2])` shoul also work but seems less readable.

File: vit_vqgan/modeling_vit_vqgan.py [hunk 3]
```diff
@@ -470,14 +465,17 @@ def setup(self):
             self.to_patches = FeedForwardLayer(
                 dim1=self.config.intermediate_size,
                 dim2=input_dim,
-                activation="tanh",
+                activation=self.config.hidden_act,
                 config=self.config,
                 dtype=self.dtype,
             )
+        self.encoder_ln = nn.LayerNorm(epsilon=self.config.layer_norm_eps, dtype=self.dtype, use_scale=False)
```
Comment (patil-suraj) [inline vit_vqgan/modeling_vit_vqgan.py:469]: We are calling `self.encoder`, in multiple methods (`self.encode` and `__call__`).
Let's move the layernorm in the encoder class, so it always get's applied. Right now it's missing from `__call__`.

File: vit_vqgan/modeling_vit_vqgan.py [hunk 4]
```diff
@@ -470,14 +465,17 @@ def setup(self):
...
 
     def encode(self, pixel_values, deterministic: bool = True):
         hidden_states = self.encoder(pixel_values, deterministic=deterministic)
-        hidden_states = self.factor_in(hidden_states, deterministic=deterministic)
+        hidden_states = ACT2FN[self.hidden_act](hidden_states)
```
Comment (patil-suraj) [inline vit_vqgan/modeling_vit_vqgan.py:473]: Not sure if we need activation after the encodr. Any particualar reason ?

File: vit_vqgan/modeling_vit_vqgan.py [hunk 5]
```diff
@@ -299,41 +299,30 @@ class VitEncoder(nn.Module):
...
+                kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+            )(hidden_states)
+        assert hidden_states.ndim == 3
 
-    def __call__(self, pixel_values, deterministic: bool = True):
-        patches = self.to_patches(pixel_values)
-        hidden_states = self.embed(patches)
-        hidden_states = hidden_states + self.position_embedding(self.position_ids)
-        hidden_states = self.transformer(hidden_states, deterministic=deterministic)
+        hidden_states = nn.Dropout(rate=self.config.dropout)(hidden_states, deterministic=deterministic)
```
Comment (patil-suraj) [inline vit_vqgan/modeling_vit_vqgan.py:324]: The `position_embeedings` layer is missing, we should add pos embeds before this dropout.

**Review by patil-suraj — COMMENTED** [review]
Looks good! Just left some comments :)

File: vit_vqgan/modeling_vit_vqgan.py (hunk 5 above)
Comment (borisdayma) [reply vit_vqgan/modeling_vit_vqgan.py:324]: You're right, I forgot to put it back when I was refactoring

File: vit_vqgan/modeling_vit_vqgan.py (hunk 2 above)
Comment (borisdayma) [reply vit_vqgan/modeling_vit_vqgan.py]: Ok, I use `hidden_states` because I'm cheating and don't have to think about the right dimension but giving expected dimensions can help prevent silent bugs

File: vit_vqgan/modeling_vit_vqgan.py (hunk 3 above)
Comment (borisdayma) [reply vit_vqgan/modeling_vit_vqgan.py:469]: Should we also move `factor_in` in encoder (and `factor_out` in decoder)?
I think that the `LayerNorm` should be applied after the projection (that way it also does not need scaling)

File: vit_vqgan/modeling_vit_vqgan.py (hunk 4 above)
Comment (borisdayma) [reply vit_vqgan/modeling_vit_vqgan.py:473]: I typically don't like 2 consecutive dense layers as they could be replaced with 1 dense layer (and less total weights) but maybe I'm just picky.
I can remove it if you want.

**Review by borisdayma — COMMENTED** [review]
I implemented the changes, mainly have comments left about moving the layernorm and the activation before projection.

File: vit_vqgan/modeling_vit_vqgan.py (hunk 3 above)
Comment (patil-suraj) [reply vit_vqgan/modeling_vit_vqgan.py:469]: I see. So let's leave it here then.

Comment (patil-suraj) [reply vit_vqgan/modeling_vit_vqgan.py:469]: Not sure about the layernorm placement. I think either should work, but let's go with this.

Comment (patil-suraj) [reply vit_vqgan/modeling_vit_vqgan.py:469]: (in the conv vae normalization happens before  the projection.)

File: vit_vqgan/modeling_vit_vqgan.py (hunk 4 above)
Comment (patil-suraj) [reply vit_vqgan/modeling_vit_vqgan.py:473]: fine with me.

**Review by patil-suraj — APPROVED** [review]
All looks good now, merging!

**Comment by borisdayma** [comment]
Oh I didn't address the layernorm comment (it's not applied in all the required functions, need to be in both __call__ and encode).
